### PR TITLE
docs: rewrite user guide, reorganize docs

### LINKED ISSUE #150: Reuse summary across readme and index
Both README.md and index.rst have the same summary of decent-bench. It would be better to put this summary in a separate file and then dynamically reuse it in readme and index to keep them in sync.

### LINKED ISSUE #151: Doc-link tests should use PR build
We have tests checking that doc-links logged to the user are valid. However, the links tested are those of the published documentation. This means that a PR making them invalid will pass, and only the next PR will fail. Hence, we should update the tests to use the links of the PR doc-build created by readthedocs.

### LINKED ISSUE #152: Tests for user guide examples
We should have tests making sure that the examples in the user guide pass mypy. This reduces the risk of updating the code but not the user guide. Therefore, a requirement is that the tests should not use copy-paste duplicates of the ones in the user guide, the examples should only be defined in a single place.

### LINKED ISSUE #254: Restructure and enhance docs (esp. user guide)
# Main tasks:

1. remove "background" page, and incorporate that discussion at the top of the user guide
2. make "advanced developer guide" a sub-page of "developer guide"
3. add biblatex citations to the docs
4. restructure user guide
5. add tests for examples, see #152 

# Addressing 3.:

- use `sphinxcontrib-bibtex` (see #159)
- then add a reference to each algorithm we provide

# Addressing 4.:

- create a more hierarchical structure, with sub-pages
- getting started page (with figures of p2p and fed networks, see #167)
- discuss the benchmarking pipeline: BenchmarkProblem -> benchmark -> BenchmarkResult -> compute_metrics -> MetricsResult -> display_metrics
- then a page on how to customize each of the components (costs, algorithm, schemes) (see #160)
- add a "philosophy" page, where we explain different design choices:
    - discussion on asynchrony in #253
    - recommended order of algorithm parameters #251 
- add a discussion of the examples in "examples" folder
- clarify which algorithms are general purpose and which are learning only (i.e. they need an `EmpiricalRiskCost`) and show how to do type-checking for the learning-only algorithms
    - a nice way to do this would be by tagging the algorithms, and then having a way to display them by tags (e.g. federated, p2p, gradient, admm, learning-only, variance-reduction, etc)

### PR DESCRIPTION
**Pull request by nicola-bastianello**
this PR completely rewrites the docs and user guide, updating it to the current state of the API, and organizing it for better readability

- the docs now contain "user guide", "API reference", "developer"
- the home page of the docs now contains the contributors list, installation info; the content is reused from README.md
- the user guide includes the pages "introduction" (with discussion of decentralized settings in general), "benchmarking" (for the guide on how to use the benchmark pipeline), "customizing" (on how to customize networks, algorithms, etc), and "utils"

other improvements:

- the bib references now point to the published version of papers, and algorithms without a reference now have it
- the extension for printing lists of tagged classes has been refactored: it now prints an inline list of classes only, and allows intersection of several tags; also, the tags are now defined in an enum
- most of the coding examples shown in the docs are now tested by pytest
- benchmark now has `n_trials=1` by default
- removed `save_path` argument from `display_metrics`; it is better to always rely on the `CheckpointManager` for defining the save path
- `display_tables` now drops the "statistic" column if it's empty (that is, if all statistics return a single value)
- the test that was checking the docs link was removed: it is fragile, and a visual inspection of the generated docs link is enough
- the examples (mnist and nim) were fixed to align with the current api

closes #150  closes #151  closes #152  closes #254

### CODE REVIEW

**Comment by Simpag** [comment]
I'll take a look at this once the rtd builds successfully

**Comment by nicola-bastianello** [comment]
> I'll take a look at this once the rtd builds successfully

done! I also made some improvements to pyproject.toml (I will do the same for decent-array when I have the time)

File: docs/source/index.rst [hunk 1]
```diff
@@ -1,25 +1,19 @@
-.. decent_bench documentation master file, created by
-   sphinx-quickstart on Mon Sep  1 21:05:43 2025.
-   You can adapt this file completely to your liking, but it should at least
-   contain the root `toctree` directive.
-
 Welcome to decent-bench!
```
Comment (Simpag) [inline docs/source/index.rst:1]: I would capitalise and remove the "-" in Decent Bench

File: README.md [hunk 2]
```diff
@@ -5,9 +5,11 @@
 https://decent-bench.readthedocs.io/)
 [![PyPI downloads](https://img.shields.io/pypi/dm/decent-bench?label=PyPI%20downloads)](
 https://pypi.org/project/decent-bench/)
+<!-- include in index.rst below this point -->
 
-decent-bench allows you to benchmark decentralized optimization algorithms under various communication constraints,
-providing realistic algorithm comparisons in a user-friendly and highly configurable setting.
+decent-bench allows you to benchmark decentralized algorithms (both peer-to-peer and federated) under realistic
```
Comment (Simpag) [inline README.md:10]: I would capitalise Decent Bench and remove

File: README.md [hunk 3]
```diff
@@ -5,9 +5,11 @@
...
+constraints (including limited communications and computational power). The code base provides a
```
Comment (Simpag) [inline README.md]: Maybe remove this parenthesis and just use a comma: "constraints, including limited communications and computational power."

File: README.md [hunk 4]
```diff
@@ -20,3 +22,8 @@ pip install decent-bench
 ## Contributors
 decent-bench is developed by [Elias Ram](https://github.com/elramen/), [Simon Granström](https://github.com/Simpag/), [Adriana Rodriguez](https://github.com/adrianardv/) under the supervision of
```
Comment (Simpag) [inline README.md:23]: Capitalize

File: README.md [hunk 5]
```diff
@@ -20,3 +22,8 @@ pip install decent-bench
 ## Contributors
 decent-bench is developed by [Elias Ram](https://github.com/elramen/), [Simon Granström](https://github.com/Simpag/), [Adriana Rodriguez](https://github.com/adrianardv/) under the supervision of 
 [Dr. Nicola Bastianello](https://bastianello.me/).
+
+
+## Contact
+Contact [Dr. Nicola Bastianello](https://bastianello.me/) if you have any questions or would like to contribute
```
Comment (Simpag) [inline README.md]: I would rephrase this so that they can start contributing without contacting you, but if they would like to get some directions they can contact you. The github issues should be tagged as "in progress" or something similar so other people don't work on features the core team is already working on

File: docs/source/user/index.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
+User Guide
+==========
+
```
Comment (Simpag) [inline docs/source/user/index.rst:3]: Maybe add some text here as a quick introduction to the subsections. Or have the introduction here but have the toc-tree at the top

File: docs/source/user/introduction.rst [hunk 7]
```diff
@@ -0,0 +1,97 @@
+.. _introduction:
+
+Introduction
+============
+
+The success of modern artificial intelligence (AI) is fueled by access to massive datasets and efficient training
+methods. However, in many use cases it is not possible to apply centralized training, for example because of
```
Comment (Simpag) [inline docs/source/user/introduction.rst]: Maybe change "for example because of" to "for example due to"?

File: docs/source/user/introduction.rst [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+Why decent-bench
+------------------
+The objective of decent-bench (for *decentralized-benchmarking*) is to support the design of decentralized algorithms that can be deployed in a wide range
+of challenging scenarios. In particular, decent-bench offers the following functionalities:
+
+* **Simulating deployment scenarios**: decent-bench allows to define realistic scenarios in which to test decentralized algorithms (both federated and peer-to-peer); this includes simulating unreliable communications and limited/heterogeneous computational power.
+* **Benchmarking**: decent-bench defines a simple benchmarking pipeline to test and compare different decentralized algorithms in the same scenario. Additionally, it provides a library of state-of-the-art algorithms ready to use.
+* **Reproducibility**: as part of the benchmarking pipeline, decent-bench allows setting random seeds to ensure that results are easily reproducible.
+* **Paper-ready results**: decent-bench provides the results of a benchmark run in a format that can be directly pasted into papers (with several customization options). The raw results are also available as ``pandas.DataFrame`` for easy inspection.
```
Comment (Simpag) [inline docs/source/user/introduction.rst]: I'm not a fan of the phrase "pasted". Maybe change it to inserted, included or something similar

File: docs/source/user/customizing/index.rst [hunk 9]
```diff
@@ -0,0 +1,29 @@
+Customizing
+===========
+
+Recalling the simple example below, which was discussed on :doc:`this page </user/benchmarking/running>`, we can highlight several
```
Comment (Simpag) [inline docs/source/user/customizing/index.rst]: Don't write "this page", have the title of the page otherwise its a bit confusing as to where you are taking the reader

File: docs/source/user/benchmarking/running.rst [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+In both examples, we generate a linear regression problem
+
+.. literalinclude:: ../../../examples/basic_p2p_example.py
+    :language: python
+    :lines: 15
+
+characterized by the local costs
+:math:`f_i(x_i) = \frac{1}{m_i} \sum_{h = 1}^{m_i} \frac{1}{2} (a_i^h x_i - b_i^h)^2` with feature vectors
+:math:`a_i^h \in \mathbb{R}^{1 \times n}` and targets :math:`b_i^h \in \mathbb{R}`. Since the costs are quadratic, the
+``create_regression_problem`` utility also computes the optimal solution ``x_optimal``.
```
Comment (Simpag) [inline docs/source/user/benchmarking/running.rst]: Explain why there's a "_" in the method return types

File: docs/source/user/benchmarking/customizing.rst [hunk 11]
```diff
@@ -0,0 +1,229 @@
+Customizing benchmark settings
+------------------------------
+
+Storing results
+^^^^^^^^^^^^^^^
+A first important tool for benchmarking is the :class:`~decent_bench.utils.checkpoint_manager.CheckpointManager`, which
+is instantiated in a folder location (folder ``results`` in the example below).
```
Comment (Simpag) [inline docs/source/user/benchmarking/customizing.rst]: "which is instantiated in a folder location" sounds a bit odd, instead write something like "which defines the result destination"

File: docs/source/user/benchmarking/customizing.rst [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+arguments are:
+
+* ``algorithms``: defines which algorithms should be tested. Each algorithm is an :class:`~decent_bench.algorithms.Algorithm` object initialized with the number of iterations to run (``iterations``) and required hyperparameters.
+* ``benchmark_problem``: instance of :class:`~decent_bench.benchmark.BenchmarkProblem` which defines the problem.
+* ``n_trials``: if the benchmark setup (network and/or algorithms) have stochastic features, running several trials is necessary. This is possible by setting the ``n_trials`` argument. See more in :ref:`this section <reproducibility>`.
+* ``max_processes``: allows to set the number of threads that should be used to run the simulations.
+* ``runtime_metrics``: if the benchmark run is very long, it can be useful to monitor its progress (beyond the progress bar that is displayed by default). This can be done by plotting :class:`~decent_bench.metrics.RuntimeMetric`, which they are simple and computationally cheap performance metric displayed in a plot that evolves as the simulations run. The available runtime metrics are: :tagged:`runtime metric`.
```
Comment (Simpag) [inline docs/source/user/benchmarking/customizing.rst]: "which they are simple" and extra "they" here, "cheap performance metric displayed" should also be "cheap performance metrics displayed"

File: docs/source/user/customizing/metric.rst [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+be displayed during execution of :func:`~decent_bench.benchmark.benchmark` (especially for very time-consuming
+simulations).
+
+decent-bench already offers a wide number of :class:`~decent_bench.metrics.Metric` objects in
+:mod:`~decent_bench.metrics.metric_library`, and the runtime metrics :tagged:`runtime metric`. If more metrics are
+needed, they can be straighforwardly implemented as subclasses of :class:`~decent_bench.metrics.Metric` and
+:class:`~decent_bench.metrics.RuntimeMetric`, providing a concrete implementation of the ``compute`` method.
+For :class:`~decent_bench.metrics.Metric`, the ``is_available`` method can also be implemented, which signals when
+a metric can be computed (*e.g.* :class:`~decent_bench.metrics.metric_library.XError` is only available
+when ``x_optimal`` for the benchmark problem was computed).
```
Comment (Simpag) [inline docs/source/user/customizing/metric.rst:15]: Metrics page could make use of an example to show how to create a new metric

File: docs/source/developer/index.rst [hunk 14]
```diff
@@ -0,0 +1,9 @@
+Developer guide
+===============
+
```
Comment (Simpag) [inline docs/source/developer/index.rst:3]: same thing as the user guide, the empty page is a bit boring

File: docs/source/developer/advanced_algorithms.rst [hunk 15]
```diff
@@ -1,5 +1,5 @@
-PyTorch Optimizer Integration
-=============================
+Advanced Algorithm Implementation
```
Comment (Simpag) [inline docs/source/developer/advanced_algorithms.rst]: maybe this page should be under user-guide instead of developer?

**Review by Simpag — COMMENTED** [review]
Overall good, some parts will need a rewrite once it is converted to using decent-array, basically all of the advanced developer guide would be removed, not sure how much is still valid for decent-array even. General feedback:
1. Remove "this page". Its a bit confusing as to what it is referencing and would be easier to read if it said the page title so the user doesnt have to click it to find out
2. I think it looks nicer to capitalise "Decent Bench" instead of writing "decent-bench". There are multiple files where this is written

File: docs/source/index.rst (hunk 1 above)
Comment (nicola-bastianello) [reply docs/source/index.rst:1]: I personally like decent-bench better since it's the name of the package

File: docs/source/developer/advanced_algorithms.rst (hunk 15 above)
Comment (nicola-bastianello) [reply docs/source/developer/advanced_algorithms.rst]: I debated that, but I think I prefer it in developer because it is not something normal users will need to care about. I didn't want the user guide to bloat too much

File: docs/source/user/customizing/metric.rst (hunk 13 above)
Comment (nicola-bastianello) [reply docs/source/user/customizing/metric.rst:15]: I decided against it, because I prefer normal users to just take metrics for granted, since we have quite an extensive library already

**Comment by nicola-bastianello** [comment]
> Overall good, some parts will need a rewrite once it is converted to using decent-array, basically all of the advanced developer guide would be removed, not sure how much is still valid for decent-array even

yes definitely. I'll adapt it to the refactored code afterwards

**Comment by nicola-bastianello** [comment]
test is green on my mac. I'm inclined to merge as is and look into this later

File: docs/source/index.rst (hunk 1 above)
Comment (Simpag) [reply docs/source/index.rst:1]: It does make sense from that perspective. Some packages use the package name while others use their full name in documentation, its fine either way imo

File: docs/source/developer/advanced_algorithms.rst (hunk 15 above)
Comment (Simpag) [reply docs/source/developer/advanced_algorithms.rst]: Then maybe have it in the documentation for the pytorch cost? I dont think it fits that well in the developer docs

**Comment by Simpag** [comment]
> test is green on my mac. I'm inclined to merge as is and look into this later

The errors are pretty far off, I wonder how that could be the case? Might want to look into how the "true" value is calculated, maybe its calculated using much higher precision? But very strange that it fails now.

Edit:
It could be because of pytorch 2.13, it was just released. Might be worth updating to 2.13 in your local env to see if that broke it. Then we'd have to lock torch down to 2.12 until they fix it in 2.13

**Comment by nicola-bastianello** [comment]
> > test is green on my mac. I'm inclined to merge as is and look into this later
> 
> The errors are pretty far off, I wonder how that could be the case? Might want to look into how the "true" value is calculated, maybe its calculated using much higher precision? But very strange that it fails now.
> 
> Edit: It could be because of pytorch 2.13, it was just released. Might be worth updating to 2.13 in your local env to see if that broke it. Then we'd have to lock torch down to 2.12 until they fix it in 2.13

that doesn't seem to be the problem; I pinned 2.13 and the test is still green. this might have to do with the version of github actions we are using. dependabot was suggesting updating that. I would merge as is and then see if updating the github actions fixes it

File: docs/source/developer/advanced_algorithms.rst (hunk 15 above)
Comment (nicola-bastianello) [reply docs/source/developer/advanced_algorithms.rst]: I think it would be confusing there.. you were right, it's best to place it in the user guide. I'll move it to the end of customizing/algorithms

**Comment by Simpag** [comment]
Im not sure whats going on with the tests, maybe try to merge the updates and rerun the test? It passes on my machine with a fresh env too... Imo its fine to merge but would be good to fix this so other commits doesn’t get false negatives

**Comment by nicola-bastianello** [comment]
> Im not sure whats going on with the tests, maybe try to merge the updates and rerun the test? It passes on my machine with a fresh env too... Imo its fine to merge but would be good to fix this so other commits doesn’t get false negatives

I incorporated dependabot's suggestions (here because otherwise I think there might have been some conflict). this didn'd fix. I tried getting more info on the set-up used in ci, and it looks like it was using macos 24. so I switched to macos-26, but it still fails, although the resulting setup is very close to the one I have locally. there seems to be several reports about inconsistencies in torch on mps, although some of them are solved in recent releases (pinning torch to latest didn't fix though)

I don't see much else we can do currently other than treating macos+pytorch tests as non-blocking. I'll merge as is for now
